packages/backend/server/src/core/doc-renderer/event.ts

```diff
@@ -0,0 +1,29 @@
+import { FactoryProvider, Injectable } from '@nestjs/common';
+import { ModuleRef } from '@nestjs/core';
+
+import { Config, OnEvent } from '../../base';
+import { DocContentService } from './service';
+
+@Injectable()
+class DocEventsListener {
+  constructor(private readonly doc: DocContentService) {}
+
+  @OnEvent('doc.snapshot.updated')
+  async handleDocSnapshotUpdated({
+    workspaceId,
+    docId,
+  }: Events['doc.snapshot.updated']) {
+    await this.doc.markDocContentCacheStale(workspaceId, docId);
+  }
+}
+
+export const DocEventsListenerProvider: FactoryProvider = {
+  provide: DocEventsListener,
+  useFactory: (config: Config, ref: ModuleRef) => {
+    if (config.flavor.renderer) {
+      return ref.create(DocEventsListener);
```

Comment:
`ModuleRef` is too low level, it can be rewritten with `new DocEventsListener(doc)`:

```ts
export const DocEventsListenerProvider: FactoryProvider = {
  provide: DocEventsListener,
  useFactory: (config: Config, doc: DocContentService) => {
    if (config.flavor.renderer) {
      return new DocEventsListener(doc);
    }
    return;
  },
  inject: [Config, DocContentService],
};
```